### DISCUSSION TITLE
generating with mpdf but not blocking all resources

### DISCUSSION THREAD

**Discussion by carstengrimm** [mpdf/mpdf#1445] [General]
I have been an frequent user of mpdf and i am currently working with pdf files which will be generated on the fly having several pictures as well as 5000+ words each. While they generate just fine on my server the generation takes a few seconds. While generation the server is not reachable at all until it has processed everything mpdf related.

My question is, if there's anyway to handle generation, that the page is still reachable, i have already thought about outputting the file and checking it's creation time to avoid making my server unreachable at several times.

### These are mPDF and PHP versions I am using
PHP 7.1 MPDF 6.1.3

**Comment by Klap-in** [comment]
This library is just for generating the pdfs.
Around the call of mpdf library you could add some logic that let it only call the mpdf if a generated pdf doesn't exist for current version of your page. 
You could use the last modification date of your page and compare that to the your cached pdf. Or even use that date as (part of) the file name of the cached file, that makes it an unique name as well. (or store all filenames and dates in a separate file. If these properties are changed you could generate a new cached pdf. etc etc. lot of possibilities ;))

Note:
When outputting the file you can save it to file, see options of the `output()` function. http://mpdf.github.io/reference/mpdf-functions/output.html

**Comment by carstengrimm** [comment]
@Klap-in thank you for your feedback. actually that's what i am doing already.

i am comparing modification dates to either push the existing file to browser or generate it when it's not there or too old which caches the file and saves resources.

my idea was to have some sort of individual pdf to protect it from theft having for example a watermark or info in footer section. while this is already possible and it will also cache the file, every new pdf generated by someone else will eat up all resources once again making server not available for 2-3 seconds or even longer depending on the size...

i have tried using compress.php but it doesn't seem to work using the latest mpdf 6.1.3 ...
i thought there might be a solution to somewhat split up resources to get generation done but still keep the server reachable....

**Comment by ssieb** [comment]
I don't understand why generating the pdf blocks the server.  Is your web server single-threaded?  Or do you mean that specific page is not available while the pdf is generating?

**Comment by carstengrimm** [comment]
@ssieb any page is not available while generation as if the server is not reachable while processing the pdf file. 

i am not sure about single-threaded or not... i guess since it's a shared server they somewhat need to make sure i'll not draw all performance by myself ...
